File: paimon-core/src/main/java/org/apache/paimon/io/SingleFileWriter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.io;

import org.apache.paimon.data.InternalRow;
import org.apache.paimon.format.BundleFormatWriter;
import org.apache.paimon.format.FormatWriter;
import org.apache.paimon.format.FormatWriterFactory;
import org.apache.paimon.fs.AsyncPositionOutputStream;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.fs.Path;
import org.apache.paimon.fs.PositionOutputStream;
import org.apache.paimon.utils.IOUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.function.Function;

/**
 * A {@link FileWriter} to produce a single file.
 *
 * @param <T> type of records to write.
 * @param <R> type of result to produce after writing a file.
 */
public abstract class SingleFileWriter<T, R> implements FileWriter<T, R> {

    private static final Logger LOG = LoggerFactory.getLogger(SingleFileWriter.class);

    protected final FileIO fileIO;
    protected final Path path;
    private final Function<T, InternalRow> converter;

    private FormatWriter writer;
    private PositionOutputStream out;

    private long recordCount;
    protected boolean closed;

    public SingleFileWriter(
            FileIO fileIO,
            FormatWriterFactory factory,
            Path path,
            Function<T, InternalRow> converter,
            String compression,
            boolean asyncWrite) {
        this.fileIO = fileIO;
        this.path = path;
        this.converter = converter;

        try {
            out = fileIO.newOutputStream(path, false);
            if (asyncWrite) {
                out = new AsyncPositionOutputStream(out);
            }
            writer = factory.create(out, compression);
        } catch (IOException e) {
            LOG.warn(
                    "Failed to open the bulk writer, closing the output stream and throw the error.",
                    e);
            if (out != null) {
                abort();
            }
            throw new UncheckedIOException(e);
        }

        this.recordCount = 0;
        this.closed = false;
    }

    public Path path() {
        return path;
    }

    @Override
    public void write(T record) throws IOException {
        writeImpl(record);
    }

    public void writeBundle(BundleRecords bundle) throws IOException {
        if (closed) {
            throw new RuntimeException("Writer has already closed!");
        }

        try {
            if (writer instanceof BundleFormatWriter) {
                ((BundleFormatWriter) writer).writeBundle(bundle);
            } else {
                for (InternalRow row : bundle) {
                    writer.addElement(row);
                }
            }
            recordCount += bundle.rowCount();
        } catch (Throwable e) {
            LOG.warn("Exception occurs when writing file " + path + ". Cleaning up.", e);
            abort();
            throw e;
        }
    }

    protected InternalRow writeImpl(T record) throws IOException {
        if (closed) {
            throw new RuntimeException("Writer has already closed!");
        }

        try {
            InternalRow rowData = converter.apply(record);
            writer.addElement(rowData);
            recordCount++;
            return rowData;
        } catch (Throwable e) {
            LOG.warn("Exception occurs when writing file " + path + ". Cleaning up.", e);
            abort();
            throw e;
        }
    }

    @Override
    public long recordCount() {
        return recordCount;
    }

    public boolean reachTargetSize(boolean suggestedCheck, long targetSize) throws IOException {
        return writer.reachTargetSize(suggestedCheck, targetSize);
    }

    @Override
    public void abort() {
        if (writer != null) {
            IOUtils.closeQuietly(writer);
            writer = null;
        }
        if (out != null) {
            IOUtils.closeQuietly(out);
            out = null;
        }
        fileIO.deleteQuietly(path);
    }

    public AbortExecutor abortExecutor() {
        if (!closed) {
            throw new RuntimeException("Writer should be closed!");
        }

        return new AbortExecutor(fileIO, path);
    }

    @Override
    public void close() throws IOException {
        if (closed) {
            return;
        }

        if (LOG.isDebugEnabled()) {
            LOG.debug("Closing file " + path);
        }

        try {
            if (writer != null) {
                writer.close();
                writer = null;
            }
            if (out != null) {
                out.flush();
                out.close();
                out = null;
            }
        } catch (IOException e) {
            LOG.warn("Exception occurs when closing file {}. Cleaning up.", path, e);
            abort();
            throw e;
        } finally {
            closed = true;
        }
    }

    /** Abort executor to just have reference of path instead of whole writer. */
    public static class AbortExecutor {

        private final FileIO fileIO;
        private final Path path;

        private AbortExecutor(FileIO fileIO, Path path) {
            this.fileIO = fileIO;
            this.path = path;
        }

        public void abort() {
            fileIO.deleteQuietly(path);
        }
    }
}

File: paimon-python/pypaimon/write/writer/single_file_writer.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from contextlib import suppress

import pyarrow.parquet as pq

from pypaimon.common.options.core_options import CoreOptions


class SingleFileWriter:
    """Write one file incrementally; currently only Parquet is supported."""

    def __init__(self, file_io, path, schema, file_format, compression, zstd_level,
                 stats_fields, stats_collector, parquet_options=None):
        if file_format != CoreOptions.FILE_FORMAT_PARQUET:
            raise NotImplementedError(
                'SingleFileWriter only supports Parquet, got {}'.format(file_format))
        self._file_io = file_io
        self._path = path
        self._stats_fields = stats_fields
        self._stats_collector = stats_collector
        self._stream = None
        self._writer = None
        self._owns_file = False
        self._closed = False
        self.row_count = 0
        self.column_stats = {}

        kwargs = dict(parquet_options or {}, compression=compression)
        if compression.lower() == 'zstd':
            kwargs['compression_level'] = zstd_level
        try:
            self._stream = file_io.new_output_stream(path)
            self._owns_file = True
            self._writer = pq.ParquetWriter(self._stream, schema, **kwargs)
        except Exception:
            self.abort()
            raise

    def write(self, data, row_group_size=None):
        if self._closed:
            raise RuntimeError('Writer is already closed')
        try:
            kwargs = {}
            if row_group_size is not None:
                kwargs['row_group_size'] = row_group_size
            self._writer.write_table(data, **kwargs)
            self.row_count += data.num_rows
            for field in self._stats_fields:
                current = self._stats_collector(data, field.name)
                previous = self.column_stats.get(field.name)
                if previous is not None:
                    current['null_counts'] += previous['null_counts']
                    for key, choose in (
                            ('min_values', min), ('max_values', max)):
                        values = [
                            value for value in (previous[key], current[key])
                            if value is not None
                        ]
                        current[key] = choose(values) if values else None
                self.column_stats[field.name] = current
        except Exception:
            self.abort()
            raise

    def close(self):
        if self._closed:
            return
        try:
            self._writer.close()
            self._writer = None
            self._stream.close()
            self._stream = None
            self._closed = True
        except Exception:
            self.abort()
            raise

    def abort(self):
        if self._writer is not None:
            with suppress(Exception):
                self._writer.close()
            self._writer = None
        if self._stream is not None:
            with suppress(Exception):
                self._stream.close()
            self._stream = None
        self._closed = True
        if self._owns_file:
            self._file_io.delete_quietly(self._path)
            self._owns_file = False
